Refactoring firmware... this PR has morphed

rvdbreemen opened the pull request:
Mindergas depended on main loop for date change. This has now been fixed, it is now independent of the main loop. Done by using the day() instead of the Thisday value from mainloop.

rvdbreemen committed:
Testing switch off

rvdbreemen committed:
Networkstuff changes

rvdbreemen committed:
Merge branch 'master' of https://github.com/rvdbreemen/DSMRloggerAPI

rvdbreemen committed:
change mqtt debug, little less

rvdbreemen committed:
hunting a bug

rvdbreemen committed:
day change detectie aangepast in mindergas

rvdbreemen committed:
Merge remote-tracking branch 'upstream/master' into bug-statemachine-mindergas

rvdbreemen committed:
cleanup code mindergas

rvdbreemen committed:
changed default display

rvdbreemen committed:
fixes long to int and changes scope from global to local

rvdbreemen committed:
Name fixed

rvdbreemen committed:
Make sure, tenminste 1 goede telegram binnen ;-)

rvdbreemen committed:
Wait for next day, not midnight ;-)

rvdbreemen committed:
oepsie... state find and replace done!

rvdbreemen committed:
Merge remote-tracking branch 'upstream/master'

rvdbreemen committed:
Merge branch 'master' into bug-statemachine-mindergas

rvdbreemen committed:
refactoring main loop to dotasks

This refactors the main loop to the dotask structure.

rvdbreemen committed:
refactored to DECLARE_TIMERS

This commit changes the whole firmware to use the safer way of using timers.
Also simplies some of the code.

rvdbreemen committed:
typo corrected

rvdbreemen commented:
@mrWheel we begonnen met een bug in mindergas statemachine. Die is nu opgelost ;-) Je nieuwe timer.h voor het veilig declareren van timers heeft mij aangezet tot het refactoren van alle millis in de firmware naar de veilige manier van declaratie, hier en daar wordt de code zelfs simpeler daardoor en vooral beter leesbaar.

rvdbreemen committed:
Peer review identified an issue with function isValidIP

The IP is valid under certain conditions. Two more cases have been added to validation

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -373,119 +373,141 @@ void setup()
+void delayms(unsigned long delay_ms)
+{
+  DECLARE_TIMER_MS(timer, delay_ms);
+  while (!DUE(timer))
```
Waarom denk je dat dit "non-blocking is"?

rvdbreemen replied:
Het is no-blocking in de zin dat de background taken actief aangeroepen worden gedurende de wachttijd. 

Het idee is dat je dus normaliter de loop aanroept, daar zitten de achtergrond taken. 

Vervolgens als je dan een “delay” wilt (wat bijna niet in de firmware zit trouwens) kan je deze functie gebruiken om je delay te doen. En dan heb je toch je achtergrond taken. 

Je hebt gelijk dat het niet non-blocking moet heten. Maar eigenlijk delay met background taken 😉

mrWheel commented on networkStuff.h:
```diff
@@ -9,11 +9,16 @@
+    #include <ESP8266WiFi.h>  //ESP8266 Core WiFi Library         
+    #include <ESP8266WebServer.h>   // Version 1.0.0 - part of ESP8266 Core https://github.com/esp8266/Arduino
+    #include <ESP8266mDNS.h>        // part of ESP8266 Core https://github.com/esp8266/Arduino
+#else
```
Hier hebben we het over gehad ..

rvdbreemen replied:
Yep, gaat eruit wat mij betreft.

rvdbreemen replied:
Yep, gaat eruit wat mij betreft.

mrWheel commented on oledStuff.h:
```diff
@@ -79,33 +76,28 @@ void oled_Init()
+    oled.print(message.c_str()); 
 
-    if (wait > 0) 
+    while (!DUE(timerWait))
```
Ik probeer te begrijpen wat hier gebeurd .. maar kom daar niet helemaal uit..

rvdbreemen replied:
Ik kijk erna. En maak het duidelijker.

rvdbreemen replied:
Ik kijk erna. En maak het duidelijker.

mrWheel committed:
Charts aangepast voor uur, dag en maand

rvdbreemen committed:
Merge branch 'AaWdev5' into bug-statemachine-mindergas

rvdbreemen committed:
Typo correct for safeTimers

rvdbreemen committed:
Changed the working of Change_interval macro

The macro changed a tiny bit, to only reset the counter when it actually changed. And I think that is needed to make it work properly in some scenario's.

rvdbreemen committed:
Manuale merge met upstream met grafieken erin

rvdbreemen committed:
Implementing Change_Interval in the code where relevant

rvdbreemen committed:
oledstuff werkt nu ook weer

en een bugje in safeTimers opgelost

rvdbreemen committed:
flxed a timer... time to debug

rvdbreemen committed:
bug hunting...

rvdbreemen committed:
Moved mindergas globals to .h file

rvdbreemen committed:
Used improved raw serial reading to capture telegrams

Copied the API serial capture code to handleRawMode better in debug mode.  Now lets fix that last little thing, the reading and processing of regular mode telegrams.

rvdbreemen committed:
Fixed the slimmermeter loop() and processing

It works!! Refactor complete.

rvdbreemen committed:
Removed the ESP32 support

No board to test it on, so I removed the porting stuff. It is cruft for now.

rvdbreemen committed:
fixed compiler optie USE_MINDERGAS

In de codebase zatten nog slordigheden rond het gebruik van USE_MINDERGAS. Als je mindergas niet wilde compileren kreeg je foutmeldingen. Nu niet meer.

rvdbreemen committed:
Fixed the countdown timer in Mindergas

Now the timers work correctly.

rvdbreemen committed:
Fixed the MQTT send in refactor

Now MQTT sending works again.

rvdbreemen committed:
Hunting the display bug

mrWheel commented on handleTestdata.ino:
```diff
@@ -30,9 +29,10 @@ void handleTestdata()
-  if (!forceBuildRingFiles && (millis() - telegramTimer) < TELEGRAM_INTERVAL)  return;   // not yet time for new Telegram
-  telegramTimer = millis();
+
+  DECLARE_TIMER_s(telegramTimer, TELEGRAM_INTERVAL);
```
Is dit getest?
Er bestaat helemaal geen macro DECLARE_TIMER_s()

mrWheel commented on handleTestdata.ino:
```diff
@@ -30,9 +29,10 @@ void handleTestdata()
+
+  DECLARE_TIMER_s(telegramTimer, TELEGRAM_INTERVAL);
+ 
+  if (!forceBuildRingFiles && (DUE( telegramTimer))  return;   // not yet time for new Telegram
```
Ik probeer te begrijpen wat hier gebeurd .. maar snap het niet..

mrWheel commented on oledStuff.h:
```diff
@@ -24,37 +24,38 @@ void oled_Print_Msg(uint8_t, String, uint16_t);
 
 uint8_t     lineHeight, charHeight;
-
+DECLARE_TIMER_MIN(oledSleepTimer, 10);
```
.. ik snap/herken mijn eigen code niet meer.
Ik vraag mij ook sterk af of dit nog werkt.

mrWheel commented on safeTimers.h:
```diff
@@ -1,8 +1,13 @@
- * by Edgar Bonet and added CHANGE_INTERVAL() macro's
+ * I made some small changes due to the "how can I handle the millis() rollover"
+ * and added the CHANGE_TIMER macro's.
+ * by Willem aan der Wiel
```
Wie is “Willem aan der Wiel” nou weer?

En waarom is de referentie naar Edgar Bonet verdwenen?

mrWheel commented on .gitignore:
```diff
@@ -0,0 +1,10 @@
+DSMRloggerAPI.ino.generic.bin
+DSMRloggerAPI.ino.generic.bin
+*.bin
+DSMRloggerAPI.ino.generic.bin
```
Mag ik hieruit opmaken dat DSMRloggerAPI.ino.generic.bin niet in github moet komen?
Zou 1x niet genoeg zijn?

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -44,6 +44,10 @@
+#ifdef USE_MINDERGAS
+  #include "MinderGas.h"
+#endif 
+
```
Waarom bovenstaande regels niet in DSMRloggerAPI.h?

Ik wil juist al deze includes uit dit file houden zodat je altijd weet waar je moet zoeken/kijken

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -77,10 +81,10 @@ void displayStatus()
   }
   oled_Print_Msg(3, cMsg, 0);
-  msgMode++;
+  msgMode= (msgMode+1) % 5; //modular 5 = number of message displayed (hence it cycles thru the messages
```
Dit zou, om dezelfde logica als voorheen te krijgen 

msgMode= (++msgMode) % 5;

Moeten worden. Het zal zó ook werken maar je verandert iets wat werkt en je weet niet of het ergens anders misschien wel een probleem gaat geven (zal in dit geval niet zo zijn, maat toch).

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -110,7 +114,7 @@ void setup()
 
 #if defined( HAS_OLED_SSD1306 ) || defined( HAS_OLED_SH1106 )
-  oledSleepTimer = millis() + (10 * 60000); // initially 10 minutes on
+  DECLARE_TIMER_MIN(oledSleepTimer, 10);
```
Door de timer hier te definiëren kun je hem nooit meer aanpassen als de settings worden aangepast.
Beter: in API.h declareren.

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -274,10 +278,10 @@ void setup()
   Serial.println("' voor verdere debugging\r\n");
   
-//=============now test if "!PRDconvert" file exists================
+//=============now test if "convertPRD" file exists================
```
Waarom?
Het file heet !PRDconvert en de weggehaalde comment was/is dus goed!

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -274,10 +278,10 @@ void setup()
   if (DSMRfileExist("!PRDconvert", false) )
   {
-    convertPRD2RING();  // convert PRD to RING file(s)
+    convertPRD2RING();
```
Toegegeven, het commentaar is wat overbodig maar waarom haal je het weg?
Origineel was/is prima

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -380,117 +384,145 @@ void setup()
+  CHANGE_INTERVAL_MS(timerBlink, iDelay);
+  if (DUE(timerBlink))
+    blinkLEDnow();
+}
```
Ongetwijfelt leuk zo’n knipperende led, maar de enige feedback die een logger zonder oled scherm heeft is de blauwe led. Die geeft aan of er een telegram binnen is gekomen.
Let wel: het lezen en verwerken van telegrammen is de core business van de logger. Nu knippert er een led zonder enige feedback over telegrammen die al-dan-niet binnen komen

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -380,117 +384,145 @@ void setup()
-  handleMQTT();
-  handleMindergas();
+  DECLARE_TIMER_MS(timer, delay_ms);
+  CHANGE_INTERVAL_MS(timer, delay_ms);
```
Waarom een declare met daar direct achter een change?
Ik blijf dit onlogisch vinden én/óf ik snap het niet

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -380,117 +384,145 @@ void setup()
+  DECLARE_TIMER_MS(timer, delay_ms);
+  CHANGE_INTERVAL_MS(timer, delay_ms);
+  while (!DUE(timer))
+    doBackgroundTasks();
```
Hier gaat bij mij het licht helemaal uit!
Waarom is doBackgroundTask() afhankelijk van “timer”?

Wellicht dat ik de context mis omdat ik niet zie wat er vóór regel 418 en ná regel 419 staat ..

mrWheel commented on DSMRloggerAPI.ino:
```diff
@@ -380,117 +384,145 @@ void setup()
+  DECLARE_TIMER_MIN(timer10min, 10)
+  DECLARE_TIMER_MIN(timer60min, 60)
+  DECLARE_TIMER_SEC(timerTelegram, settingInterval)
+  CHANGE_INTERVAL_SEC(timerTelegram, settingInterval)
```
Ook hier snap ik de change direct na de declare niet.
Door de declare hier te doen kun je timerTelegram niet meer veranderen als de settingInterval veranderd.
Niet goed.

mrWheel commented on MQTTstuff.ino:
```diff
@@ -17,20 +17,23 @@
+  DECLARE_TIMER_MS(mqttTimer, 500);                       //every 500 ms do handle state
+  DECLARE_TIMER_SEC(mqttRetryTimer, 3);                   //backoff timer 
+  DECLARE_TIMER_MIN(timeMQTTReconnect, 10)                //try reconnecting cyclus timer
+  DECLARE_TIMER_SEC(timeMQTTPublish,  ((settingMQTTinterval == settingInterval) ? (settingMQTTinterval-1):settingMQTTinterval)); //special case, if telegram interval = mqtt interval, then mqtt interval needs to be shorter
```
Waarom moet de MQTT interval altijd kleiner zijn dan de telegram interval?
Het heeft geen zin om een MQTT boodschap te versturen als er geen nieuw telegram is..
.. en nogmaals: het is géén realtime os. Alles wordt sequentieel/achter/na elkaar afgehandeld. 
MQTT interval mag gróter zijn dan telegram interval, maar nooit kleiner!

mrWheel commented on MinderGas.h:
```diff
@@ -0,0 +1,40 @@
+**
+**  TERMS OF USE: MIT License. See bottom of file.                                                            
+***************************************************************************  
+*/
```
Liever gewoon in Mindergas.ino opnemen óf in API.h

mrWheel commented on handleSlimmeMeter.ino:
```diff
@@ -10,12 +10,96 @@
 
 #if !defined(HAS_NO_SLIMMEMETER)
+//==================================================================================
+void handleSlimmemeter(){
```
Ok, nieuwe standaard die ik gebruik is dat alle ‘{‘ en ‘}’ op een nieuwe regel en onder elkaar staan (behalve als het er tientallen zijn en het de leesbaarheid eerder tegen werkt dan beter maakt).

.. en zo niet dan toch wel een spatie vóór de ‘{‘ ..

mrWheel commented on handleSlimmeMeter.ino:
```diff
@@ -10,12 +10,96 @@
+void handleSlimmemeter(){
+  
+  //DebugTf("showRaw (%s)\r\n", showRaw ?"true":"false");
+  if (showRaw) {
```
Liever dus:

If (showRaw)
{
   ...
}

mrWheel commented on ntpStuff.ino:
```diff
@@ -76,8 +76,8 @@ time_t getNtpTime()
     sendNTPpacket(ntpServerIP);
-    uint32_t beginWait = millis();
-    while (millis() - beginWait < 1500) 
+    DECLARE_TIMER_MS(beginWait, 1500);
```
Wat is hier de toegevoegde waarde van de declare timer?
Je moet gewoon 1x maximaal 1500ms wachten of je wat binnen hebt gekregen. In dit geval vind ik de timer niet voor duidelijkheid zorgen.
Eventueel zou je er: 
   while ((int32_t)(millis() - beginWait) < 1500)
van kunnen maken..

rvdbreemen committed:
corrected .gitignore

rvdbreemen committed:
Merge remote-tracking branch 'upstream/master' into bug-statemachine-mindergas

rvdbreemen committed:
fixed final bugs

- fixed the oled display to working order
- reverted the safeTimers.h to a working edition
- removed mindergas.h
- fixed bug in handleTestdata.ino with timer macro
- fixed MQTT IPaddress conversion bug

rvdbreemen committed:
final cleanup and removed blinkled(1000)

rvdbreemen committed:
Added WAIT_FOR_SETUP in MQTT FSM handler

rvdbreemen committed:
Final fixes (for now)

rvdbreemen committed:
Improved the display off/on switching with flash

rvdbreemen committed:
removed not needed call from MQTT

rvdbreemen committed:
cleanup code

rvdbreemen committed:
Updated safeTimers MACRO name to reflect the use

The MACRO name should reflect how it works. Now it does.
